### ISSUE TITLE
Weird inference result

### ISSUE THREAD

**Issue by angelacao-hub** [HengruiTianSJTU/LesionDiffusion#8]
Dear authors,

I tried to inference with one of the demo image located in `demo/liver_0/raw.nii.gz`, however after going through the whole inference process, the resulting `inpainting_0.nii.gz` looks like random noise as illustrated below.
<p align="center">
<img width="30%" alt="inpainting_0.nii.gz" src="https://github.com/user-attachments/assets/fd31a4f5-210c-4afc-aa3c-aea5d168a5ac" />
</p>

---

The following is the process I used to run inference on the image. I was wondering whether I might have done something incorrectly.

## Preprocess
First, I copied `demo/liver_0/raw.nii.gz` and to `example/raw.nii.gz`.
<p align="center">
<img width="30%"  alt="raw.nii.gz" src="https://github.com/user-attachments/assets/5b7c2eb7-71d4-4f4d-bc73-5eebf1633567" />
</p>
Inside the `example` folder, I created `example/pre_img_list.txt` which contains

```
../example/raw.nii.gz
```

and `example/pre_attr_list.txt` which contains

```
{"organ type": "liver", "lesion type": "cyst"}
```

After setting up the `OPENAI_API_KEY`, I ran
```
python preprocess.py ../example/pre_img_list.txt exp ../example/pre_attr_list.txt
```

which generated `example/type.json` that contains
```
{
    "5": [
        "round-like",
        "",
        {
            "enhancement status": "non-contrast ct",
            "lesion location": 5,
            "shape": "round-like",
            "size": "[[220, 151, 124]] pixel",
            "density": "hypodense lesion",
            "density variations": "homogeneous",
            "surface characteristics": "well-defined margin",
            "relationship with adjacent organs": "no close relationship with surrounding organs",
            "specific features": "none",
            "cavity": "within the parenchymal organ"
        },
        [
            [[220, 151, 124]],
            [["X", "Y", "Z"]]
        ]
    ]
}
```
`example/totalseg.nii.gz` and `example/bbox.nii.gz` which look like
<p align="center">
<img width="30%"  alt="totalseg & bbox.nii.gz" src="https://github.com/user-attachments/assets/151ab420-54f4-40a0-a9f9-dfe37c34d287" />
<img width="7%" alt="Image" src="https://github.com/user-attachments/assets/92fe4e28-3eee-4974-a3bd-7f7516b1767c" />
</p>

It also generated `pipeline/exp/bbox_list.txt`
```
../example/bbox.nii.gz
```
`pipeline/exp/img_list.txt`
```
../example/raw.nii.gz
```
`pipeline/exp/seg_list.txt`
```
../example/totalseg.nii.gz
```
`pipeline/exp/type_list.txt`
```
../example/type.json
```

## LMNet
Then I downloaded the pretrained weights provided in [Google Drive](https://drive.google.com/drive/folders/1n6_eAWhsHBFBYtDxPtp_FXp_tNHhOCHi) into `pipeline` as `pipeline/LINet.ckpt`, `pipeline/LMNet.ckpt` and `pipeline/VQ.ckpt`.

I also changed the `save_dataset_path` in both `configs/diffusiomaskdiffusion_test.yaml` and `configs/diffusion/lesiondiffusion_test.yaml` into `../example` to matach my input data dir.

After setting up these files, I ran this commad following the `README.md`
```
python test.py --base configs/diffusion/maskdiffusion_test.yaml --name mask-diff-infer --gpus 0
```
This produces `example/samples_0.nii.gz` which looks like
<p align="center">
<img width="30%"  alt="samples_0.nii.gz" src="https://github.com/user-attachments/assets/60b813de-0ab3-4213-a6c6-10d734d2372f" />
<img width="30%"  alt="totalseg & samples_0.nii.gz" src="https://github.com/user-attachments/assets/a46f9323-a540-40b9-81c8-20ce5f3d1628" />
</p>
and also a process image `example/images/test/gs-000000_e-000000_b-000000.png`
<p align="center">
<img width="50%"  alt="Image" src="https://github.com/user-attachments/assets/9433cc11-a192-421d-ba32-1791415c52e8" />
</p>

## LINet
All the output prior to running the LINet looks resonable to me. For execution of LINet, I created `pipeline/exp/label_list.txt` which contains
```
../example/samples_0.nii.gz
```
I follow `README.md` and ran
```
python test.py --base configs/diffusion/lesiondiffusion_test.yaml --name inp-diff-infer --gpus 0, 1
```
This produces `example/inpainting_0.nii.gz` which I presume is the final result, but it looks like random noise as illustrated below and at the begining of this post.
<p align="center">
<img width="30%" alt="inpainting_0.nii.gz" src="https://github.com/user-attachments/assets/fd31a4f5-210c-4afc-aa3c-aea5d168a5ac" />
</p>

---

Hope that I can get some guidance on what did I do wrong with my inferencing. Thanks in advance.

**Comment by HengruiTianSJTU** [comment]
Hi @angelacao-hub ,

Thanks so much for the detailed report and the logs! It’s really helpful to see such clear information. I also appreciate your continued support—this is the second issue you’ve raised, and your feedback is invaluable to us.

Regarding this error, it seems to be related to specific hardware configurations. We primarily develop and test on **NVIDIA RTX 4090**, where the code runs smoothly. However, a colleague of mine encountered a similar issue when trying to reproduce it on an **A100**.

He managed to resolve this on the A100 by making a slight adjustment in `pipeline/ldm/models/diffusion/ddpm.py`. Specifically, in the `log_images` function of the class `InpaintingDiffusion_v2`, they changed:

```python
# Original line causing issues on some hardware
x_samples = self.decode_first_stage(z, cf, split_quantize=True)

# Modified line that worked for them
x_samples = self.first_stage_model.decode(z)
```

Since we cannot fully verify the side effects across all GPU architectures right now, we hesitate to merge this change into the main branch immediately.

Could you please try this modification on your end and let me know if it resolves the problem?

Best regards.

**Comment by angelacao-hub** [comment]
Thanks for your answer, I'm working on A6000, I've tried the modified code, the resulting `example/inpainting_0.nii.gz` looks reasonable now!

<p align="center">
<img width="30%" alt="inpainting_0.nii.gz" src="https://github.com/user-attachments/assets/bdda0937-d9ee-429b-878b-53cfde81c610" />
</p>

As the aim of the model is to add synthetic anomaly into CT image, so I also tried to visualize where the `inpainting_0.nii.gz` volume differ from the original volumn `raw.nii.gz`. But I couldn't see where the anomaly have been generated. From my perspective, the difference volume does not appear to show a clear structured pattern. This is how I generated the difference volumn with percentile of 90, 95 and 99.5.

```
diff = np.abs(vol_raw - vol_inpaint)
thres = np.percentile(diff, perc)
vol = (diff > thres).astype(np.float64)
```

and the following is the visualized result
<p align="center">
<img width="25%" alt="90 percentile" src="https://github.com/user-attachments/assets/fe70f5fd-90a6-4694-a68c-78831e02282b" />
<img width="25%" alt="95 percentile" src="https://github.com/user-attachments/assets/77784368-3fa7-486d-9524-87d086964580" />
<img width="25%" alt="99.5 percentile"src="https://github.com/user-attachments/assets/123e4361-4b30-45aa-8304-b9ff8e75dcb4" />
<img width="10%" alt="legend" src="https://github.com/user-attachments/assets/dac9bda3-571f-4685-b89f-c928a3ec447a" />
</p>

Is this kind of result expected?

**Comment by HengruiTianSJTU** [comment]
Hi @angelacao-hub,

Sorry for the delayed response. We’ve been caught up with other tasks. Thank you for your patience!

Regarding your new observation about the difference map, I have analyzed the images you provided, and here is what I suspect is happening:

### 1. Analysis of the Difference Map
From the visualization, it appears that the differences are located **outside** the target bounding box (bbox). This suggests that `LINet` might not have successfully performed the intended inpainting inside the bbox. Instead, what you are seeing in the difference map is likely the **reconstruction discrepancy from the VQ-VAE stage**, rather than the synthetic lesion generated by the diffusion model. 

### 2. Possible Causes

This pipeline is composed of several sequential stages (caption generation → mask diffusion → inpainting diffusion), so failures in intermediate conditioning can propagate downstream. A few possible reasons are:

- The LLM generates textual conditions that might occasionally be out-of-distribution (OOD) or ambiguous for the downstream diffusion models, leading to suboptimal generation.
- Numerical/environment differences across different software stacks (CUDA, PyTorch, GPU architecture, etc.) can sometimes affect stochastic sampling behavior, even if the pipeline itself runs successfully.

### 3. Next Steps & Troubleshooting
To narrow down whether this is a data-specific issue or a broader configuration problem, could you please try the following?

*   **Option A: Try other demo cases.**
    Could you try running the inference on the other demo `.nii.gz` files provided in the `demo/` folder? Or, if you have access to other datasets like MSD (Medical Segmentation Decathlon), try using samples from these dataset as mentioned in our paper. *Note: Please ensure you resize them to our standard input size of 280^3.* If other cases work well, it might be an issue specific to the features of `liver_0`.

*   **Option B: Share your workspace for debugging.**
    If the issue persists across multiple samples, could you please pack your current `example/` directory (including `raw.nii.gz`, `bbox.nii.gz`, `totalseg.nii.gz`, `type.json`, `samples_0.nii.gz` and `inpainting_0.nii.gz`) and share it via a cloud drive link (e.g., Google Drive, OneDrive)? Since GitHub issues only support images, having the actual data would allow me to reproduce your exact two-step inference process on my end and debug where the pipeline breaks down.

Thanks again for your detailed feedback and for helping us improve the robustness of the code. 

Best regards.

**Comment by angelacao-hub** [comment]
Thank you very much for your detailed analysis and suggestions. I appreciate your patience and your willingness to help debug this issue. Regarding the difference map, I agree with your observation that the changes appear to be mostly outside the target bbox. It is helpful to know that this may correspond to reconstruction discrepancy from the VQ-VAE stage rather than the intended lesion inpainting result from the diffusion model.

## Attempt with Option A

I tried running the pipeline on the other demo cases as suggested.

First, I tested `demo/Bladder_cancer/img_health.nii.gz`. Since this appears to be a bladder cancer case, I used the following entry in `pre_attr_list.txt`:

```json
{"organ type": "bladder", "lesion type": "cancer"}
```

However, `preprocess.py` failed on this demo volume with the following error:

```text
Organ bladder not found in totalseg_class dictionary.
```

I then tested the other demo image, `demo/Liver_cancer/img_health.nii.gz`. In this case, the pipeline completed successfully, but the difference map of `inpainting_0.nii.gz` still appears to look more like generation/reconstruction noise rather than a clear lesion inpainting result, as shown below:

<p align="center">
  <img width="30%" alt="difference map" src="https://github.com/user-attachments/assets/f320c673-1e5e-4ee6-a4ce-59fd36e8b23e" />
</p>

Based on these results, the issue does not seem to be limited to a single demo case on my side. Please let me know if I should try a different organ/lesion setting for the bladder case, or if there are any additional intermediate files or logs that would be helpful for debugging.

# Option B

Thank you very much for offering this option and for your willingness to help debug the issue.

I have created a [Google Drive folder](https://drive.google.com/drive/folders/1kB5WucA-h-Py5apNPYr4whd7JNRSjmy5?usp=sharing) containing the two demo cases for which I was able to successfully generate `inpainting_0.nii.gz`.

Please let me know if any additional files or information would be helpful.
Best regards.

**Comment by HengruiTianSJTU** [comment]
Hi @angelacao-hub,

We tested your `liver_0` case on our side, using your provided intermediate results directly. We did not rerun preprocessing or LMNet; for LINet we used:

- `raw.nii.gz`
- `totalseg.nii.gz`
- `type.json`
- `samples_0_ac.nii.gz`

With the released `VQ.ckpt` and `LINet.ckpt`, the LINet output on our side does not become random noise. The figure below shows one axial slice around the generated mask region:

<img width="2240" height="1120" alt="Image" src="https://github.com/user-attachments/assets/06815e76-0b78-47ca-afc6-0377dcf95c71" />

> One thing to note is that the generated mask itself is not ideal in this example, since part of it is outside the liver region. Even with this imperfect mask, LINet still performs structured inpainting in our run, so the second-stage inference does not seem to fail in the same way as your original result.

We also compared our environment with the released `requirements.txt`, and the main dependencies are aligned:

```text
Python 3.11.9
torch 2.1.0
torchvision 0.16.0
pytorch-lightning 1.7.7
numpy 1.26.3
SimpleITK 2.3.1
nibabel 5.2.1
torchio 0.19.6
monai 1.3.0
```

At this point, we have not found another clear cause from this example alone. The most useful next step may be to double-check the full software environment against the released requirements, especially the PyTorch/CUDA-related packages. If convenient, it would also be helpful to try the same LINet step on a different GPU architecture/model, or in a fresh environment with the provided requirements, to see whether the LINet inpainting failure persists.

For the bladder demo preprocessing error, the `organ type` should follow the TotalSegmentator-style names listed in `pipeline/organ_type.json`, as mentioned in the README. In this case, please use `"urinary bladder"` instead of `"bladder"`; the corresponding entry is:

```json
"104": "urinary bladder"
```

Best regards.
